### ISSUE TITLE
Compile using CRTBNDRPG will not create an object

### ISSUE THREAD

**Issue by Alfmesiter** [codefori/vscode-ibmi#1518]
/home/bmiller: /QOpenSys/usr/bin/qsh
liblist -d QGPL QTEMP TESTLIB TESTPLIB WEUPD WEP ESEND WESQL SEQUEL TAATOOL KULIB KOLIB #DTAWHSWE LIBFTP LIBHTTP JOURNLIB && liblist -c BMILLER && liblist -a JOURNLIB LIBHTTP LIBFTP #DTAWHSWE KOLIB KULIB TAATOOL SEQUEL WESQL ESEND WEP WEUPD TESTPLIB TESTLIB QTEMP QGPL && system  "CRTBNDRPG PGM(BMILLER/RULESR) SRCFILE(BMILLER/RPGLE) OPTION(*EVENTF) DBGVIEW(*ALL) TGTRLS(*CURRENT)"; if [[ $? -ne 0 ]]; then exit 1; fi
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/bmiller: system "QSYS/CPYTOIMPF FROMFILE(BMILLER/EVFEVENT RULESR) TOSTMF('/tmp/vscodetemp-O_jcuJmQ6z') MBROPT(*REPLACE) STMFCCSID(1208) RCDDLM(*CRLF) DTAFMT(*DLM) RMVBLANK(*TRAILING) ADDCOLNAM(*SQL) FLDDLM(',') DECPNT(*PERIOD)"
{
    "code": 255,
    "signal": null,
    "stdout": "",
    "stderr": "CPF2803: To-file EVFEVENT in BMILLER not allowed.\nCPF2817: Copy command ended because of error."
}

**Comment by chrjorgensen** [comment]
@Alfmesiter Seems like there's a problem with your EVFEVENT file in library BMILLER - maybe an authority or CCSID issue?

Check if your command works in a 5250 session:

```
CRTBNDRPG PGM(BMILLER/RULESR) SRCFILE(BMILLER/RPGLE) OPTION(*EVENTF) DBGVIEW(*ALL) TGTRLS(*CURRENT)
```
If no problem found here, delete the BMILLER/EVFEVENT file and try this operation in Code for IBM i again (to have the EVFEVENT file automatically created).

**Comment by chrjorgensen** [comment]
@Alfmesiter Please verify your library list in Code for IBM i by disconnect, right-click the system and select `Connect and Reload Server Settings`.

This will check all your server settings, including validating every library in your list and offer to remove any libraries, that does not exist.

Having a "dead" library in the list will prohibit any compiles from succeeding...

**Comment by Alfmesiter** [comment]
I reloaded the server settings. In the message, it stated that libraries QGPL and QTEMP, along with test data libraries, didn't exist. I accepted the option to remove them from the list.

I attempted to compile an object and received the following message:
/home/bmiller: system "QSYS/CPYTOIMPF FROMFILE(BMILLER/EVFEVENT RESETR) TOSTMF('/tmp/vscodetemp-O_t2OwYi4c') MBROPT(*REPLACE) STMFCCSID(1208) RCDDLM(*CRLF) DTAFMT(*DLM) RMVBLANK(*TRAILING) ADDCOLNAM(*SQL) FLDDLM(',') DECPNT(*PERIOD)"
{
    "code": 255,
    "signal": null,
    "stdout": "",
    "stderr": "CPF2803: To-file EVFEVENT in BMILLER not allowed.\nCPF2817: Copy command ended because of error."
}

I deleted the EVFEVENT file in BMILLER, and attempted another compile:
/home/bmiller: /QOpenSys/usr/bin/qsh
liblist -d QGPL QTEMP TESTLIB TESTPLIB WEUPD WEP ESEND WESQL SEQUEL TAATOOL KULIB KOLIB #DTAWHSWE LIBFTP LIBHTTP JOURNLIB && liblist -c BMILLER && liblist -a  && system  "CRTBNDRPG PGM(BMILLER/RESETR) SRCFILE(BMILLER/RPGLE) OPTION(*EVENTF) DBGVIEW(*ALL) TGTRLS(*CURRENT)"; if [[ $? -ne 0 ]]; then exit 1; fi
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/bmiller: system "QSYS/CPYTOIMPF FROMFILE(BMILLER/EVFEVENT RESETR) TOSTMF('/tmp/vscodetemp-O_t2OwYi4c') MBROPT(*REPLACE) STMFCCSID(1208) RCDDLM(*CRLF) DTAFMT(*DLM) RMVBLANK(*TRAILING) ADDCOLNAM(*SQL) FLDDLM(',') DECPNT(*PERIOD)"
{
    "code": 255,
    "signal": null,
    "stdout": "",
    "stderr": "CPF2817: Copy command ended because of error."
}

It didn't create the object of the EVFEVENT file in BMILLER

**Comment by Markj0616** [comment]
@Alfmesiter were you able to identify and fix this issue? I have tried everything listed above and am still having the same issue you were having. The compile does not return any errors, but I'm getting no object on the IBM i ... also i am not getting any spool file ... maybe a clue?

**Comment by Alfmesiter** [comment]
I tried all the solutions that were given to me by tech support, and it didn't resolve the issue.

Sorry I couldn't offer any additional help

**Comment by Markj0616** [comment]
not sure if it's connected, but we have found that when i connect to my system, a job called QZLSFILE is throwing hundreds of Not authorized to object errors like it's trying to do something with a ton of objects. I was going to try to delete my connection and re-create it with different settings, but i can't figure out how to delete a connection

**Comment by Alfmesiter** [comment]
Connect to the server, then right click on the connection to get the pop up window. Click the delete connection


***@***.***

**Comment by chrjorgensen** [comment]
> we have found that when i connect to my system, a job called QZLSFILE is throwing hundreds of Not authorized to object errors like it's trying to do something with a ton of objects.

@Markj0616 The job QZLSFILE is the NetServer job, which serves the SMB shares to the SMB (Windows) clients, and this is not used at all by Code for IBM i. Your connection from Code for IBM i solely uses the SSH server running on IBM i, probably in the QUSRWRK subsystem in a job called QP0ZSPWP.

Please try running the `CRTBNDRPG` command in a 5250 session - and note if an EVFEVENT file and member is created in the program object library.

**Comment by chrjorgensen** [comment]
> I deleted the EVFEVENT file in BMILLER, and attempted another compile:
/home/bmiller: /QOpenSys/usr/bin/qsh
liblist -d QGPL QTEMP TESTLIB TESTPLIB WEUPD WEP ESEND WESQL SEQUEL TAATOOL KULIB KOLIB #DTAWHSWE LIBFTP LIBHTTP JOURNLIB && liblist -c BMILLER && liblist -a  && system  "CRTBNDRPG PGM(BMILLER/RESETR) SRCFILE(BMILLER/RPGLE) OPTION(*EVENTF) DBGVIEW(*ALL) TGTRLS(*CURRENT)"; if [[ $? -ne 0 ]]; then exit 1; fi

@Alfmesiter Please check you library list - the above command states `liblist -c BMILLER && liblist -a &&` which seems to indicate that Code for IBM i runs the command with only `BMILLER` as current library and an empty user portion of the library list.

If Code for IBM i complains about libraries in your library list, check that no libraries contains variant characters in their name or your CCSID value is correct. Read more here: https://codefori.github.io/docs/#/pages/tips/ccsid

**Comment by Alfmesiter** [comment]
Other than defining the libraries in the User Library List(below), is there another are the list must be defined for compiling? Thanks!

***@***.***

**Comment by worksofliam** [comment]
@Alfmesiter Images cannot be shared over email. Please share them directly in the issue on GitHub.

Otherwise, no. There is only one place to specify the library list, which is the User Library List view.

**Comment by Markj0616** [comment]
When I copy the CRTBNDRPG from VSCode and run it from the 5250 screen, it creates both the EVFEVENT and the program member in my library


From: Christian Jorgensen ***@***.***>
Sent: Saturday, November 4, 2023 5:24 PM
To: codefori/vscode-ibmi ***@***.***>
Cc: Mark Johnson ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [codefori/vscode-ibmi] Compile using CRTBNDRPG will not create an object (Issue #1518)

You don't often get email from ***@***.*** Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>

we have found that when i connect to my system, a job called QZLSFILE is throwing hundreds of Not authorized to object errors like it's trying to do something with a ton of objects.

@Markj0616<https://github.com/Markj0616> The job QZLSFILE is the NetServer job, which serves the SMB shares to the SMB (Windows) clients, and this is not used at all by Code for IBM i. Your connection from Code for IBM i solely uses the SSH server running on IBM i, probably in the QUSRWRK subsystem in a job called QP0ZSPWP.

Please try running the CRTBNDRPG command in a 5250 session - and note if an EVFEVENT file and member is created in the program object library.

-

**Comment by Alfmesiter** [comment]
It was just a screen shot of my lib list in VS

**Comment by bpmcpherson** [comment]
@Alfmesiter I don't know if this is your problem, but I was experiencing something similar. Through much trial and error, I discovered that the underlying command run through the PASE shell is the source of the failure.

If extract the command similar to the one below which is generated by the extension and run it, see if you get any errors such as existing libraries.

liblist -d MYLIBCMD MYLIB LEGACY QGPL PRJPRDDTA PRJPRDOBJ && liblist -c PRJPRDOBJ && liblist -a BMCPHERSON MYLIB PRJPRDOBJ QGPL QTEMP && system  "CRTSQLRPGI OBJ(PRJPRDOBJ/RPG101) SRCFILE(PRJPRDOBJ/QRPGLESRC) SRCMBR(RPG101) OPTION(*EVENTF) DBGVIEW(*SOURCE) OBJTYPE(*PGM) TGTRLS(*CURRENT)"; if [[ $? -ne 0 ]]; then exit 1; fi

Next, use the "liblist" command to see if the libraries you are expecting in your library list are showing up. My issue was that I required a library (mylib) as it contained a copybook, but it was not being re-introduced by the command. Thus, my compilation was failing when trying to apply the copybook. If discovered this by isolating the following command and running it, again in a PASE shell.

system  "CRTSQLRPGI OBJ(PRJPRDOBJ/RPG101) SRCFILE(PRJPRDOBJ/QRPGLESRC) SRCMBR(RPG101) OPTION(*EVENTF) DBGVIEW(*SOURCE) OBJTYPE(*PGM) TGTRLS(*CURRENT)"

I am still fiddling with settings to see if I can make the entire command work without issue, but it's the discovery I have made. 

Hope it helps. :)

**Comment by chrjorgensen** [comment]
@Alfmesiter I think I've found the problem causing your compile to fail: You have a library in your library list with a name starting with `#` - this causes the shell command to become like `liblist ... #DTAWHSWE LIBFTP ...`, and ` #` makes the remaining characters a comment in the command!

This is normal shell behavior - to allow comments after command by starting the comment with a `#` - and unfortunately not something handled in the code.

I've created a PR #1673 to fix this problem - when approved and merged, your problem should also be fixed.
